### PR TITLE
Static and ordered album or "images order" aware tags

### PR DESCRIPTION
**Pull request by phweyland**
Attempt to solve #5018 
Main changes
- add album attribute to tag
- add album filter entry in collect module
- add image order selection to album/tag info (highest 16 bits of flags)
- add position to tagged_images
- remove caption & description fields from images table
- position update on tag attachment
- position change on drag & drop for albums
- legacy params

For the user:
An album is a tag which has the "album" attribute (as private or category).

image

A new entry in collect module allows to filter the albums of the db. Unlike tag filter, the user can only see the images of that album, whatever its place in the hierarchy.

image

When the first collect filter is "album", any order change is saved with that specific album, including the custom sort image order.

If the first collect filter is not an album, the order changes are applied on the current collection, just as before.

### CODE REVIEW

File: src/common/collection.h [hunk 1]
```diff
@@ -76,15 +76,18 @@ typedef enum dt_collection_sort_t
   DT_COLLECTION_SORT_TITLE,
   DT_COLLECTION_SORT_DESCRIPTION,
   DT_COLLECTION_SORT_ASPECT_RATIO,
-  DT_COLLECTION_SORT_SHUFFLE
+  DT_COLLECTION_SORT_SHUFFLE,
```
Comment (TurboGit) [inline src/common/collection.h:79]: this is unwanted

File: src/common/database.c [hunk 2]
```diff
@@ -1611,6 +1611,56 @@ static int _upgrade_library_schema_step(dt_database_t *db, int version)
...
+
+    TRY_EXEC("INSERT INTO main.i SELECT id, group_id, film_id, width, height, filename, maker, model,"
+             " lens, exposure, aperture, iso, focal_length, focus_distance, datetime_taken, flags,"
+             " output_width, output_height, crop, raw_parameters, raw_denoise_threshold,"
+             " raw_auto_bright_threshold, raw_black, raw_maximum, license, sha1sum,"
+             " orientation, histogram, lightmap, longitude, latitude, altitude, color_matrix, colorspace, version,"
+             " max_version, write_timestamp, history_end, position, aspect_ratio, exposure_bias,"
+             " import_timestamp, change_timestamp, export_timestamp, print_timestamp "
+             "FROM main.images",
+             "[init] can't populate table i\n");
```
Comment (TurboGit) [inline src/common/database.c:1655]: when doing that the index are discarded. You need to recreate them:
- image_position_index   (note that there is no s on this one)
- images_filename_index
- images_film_id_index
- images_group_id_index

File: src/common/image.c [hunk 3]
```diff
@@ -865,8 +865,13 @@ int32_t dt_image_duplicate_with_version(const int32_t imgid, const int32_t newve
     sqlite3_finalize(stmt);
 
     DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
-                                "INSERT INTO main.tagged_images (imgid, tagid)"
-                                "  SELECT ?1, tagid FROM main.tagged_images WHERE imgid = ?2",
+                                "INSERT INTO main.tagged_images (imgid, tagid, position)"
+                                "  SELECT ?1, tagid, "
+                                "        (SELECT (IFNULL(MAX(position),0) & (4294967295 << 32))"
```
Comment (TurboGit) [inline src/common/image.c:870]: 4294967295 << 32  ???? Isn't this expected to be "2 << 32" ?

The above is a 64 bits integer where the 32 first bits (high order) are set to 1. Maybe expected, but I prefer to ask, not clear reading the code.

File: src/common/image.c [hunk 4]
```diff
@@ -1210,9 +1215,9 @@ static uint32_t dt_image_import_internal(const int32_t film_id, const char *file
   //insert a v0 record (which may be updated later if no v0 xmp exists)
   DT_DEBUG_SQLITE3_PREPARE_V2
     (dt_database_get(darktable.db),
-     "INSERT INTO main.images (id, film_id, filename, caption, description, license, sha1sum, flags, version, "
+     "INSERT INTO main.images (id, film_id, filename, license, sha1sum, flags, version, "
      "                         max_version, history_end, position, import_timestamp)"
-     " SELECT NULL, ?1, ?2, '', '', '', '', ?3, 0, 0, 0, (IFNULL(MAX(position),0) & (4294967295 << 32))  + (1 << 32), ?4 "
+     " SELECT NULL, ?1, ?2, '', '', ?3, 0, 0, 0, (IFNULL(MAX(position),0) & (4294967295 << 32))  + (1 << 32), ?4 "
```
Comment (TurboGit) [inline src/common/image.c:1220]: likewise ?

File: src/common/image.c [hunk 5]
```diff
@@ -1811,9 +1816,12 @@ int32_t dt_image_copy_rename(const int32_t imgid, const int32_t filmid, const gc
         sqlite3_step(stmt);
         sqlite3_finalize(stmt);
         DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
-                                    "INSERT INTO main.tagged_images (imgid, tagid)"
-                                    " SELECT ?1, tagid"
-                                    " FROM main.tagged_images"
+                                    "INSERT INTO main.tagged_images (imgid, tagid, position)"
+                                    " SELECT ?1, tagid, "
+                                    "        (SELECT (IFNULL(MAX(position),0) & (4294967295 << 32))"
```
Comment (TurboGit) [inline src/common/image.c:1821]: likewise.

File: src/common/tags.c [hunk 6]
```diff
@@ -63,7 +63,13 @@ static gchar *_get_tb_added_tag_string_values(const int img, GList *before, GLis
   {
     if(!g_list_find(b, a->data))
     {
-      tag_list = dt_util_dstrcat(tag_list, "(%d,%d),", GPOINTER_TO_INT(img), GPOINTER_TO_INT(a->data));
+      tag_list = dt_util_dstrcat(tag_list,
+                                 "(%d,%d,"
+                                 "  (SELECT (IFNULL(MAX(position),0) & (4294967295 << 32)) + (1 << 32)"
```
Comment (TurboGit) [inline src/common/tags.c:68]: likewise

**Comment by phweyland** [comment]
>  This could really be done with a standard tag, right? The point here is to make more visible a way to have images grouped in album and make this a first class citizen. Right ?
> 
> Now the question is do we really want this added complexity where one could use a hierarchical tag structure

Good questions... This is more a question of usage.

I started with the idea of album concept but ended up by having the sort property on all tags (just because it is easier and faster to apply it everywhere).
So, in collect module, having tag as the first filter, we could have the "album" effect for any tag with the same code.
The single drawback I see is that I find tag filter in collect module a bit heavy and slow. Having the album filter can make it a bit faster.

It would be interesting to get users' opinion here !

A more technical aspect:
Another possibility I thought about is to remove position from images tables and use only tagged_images table position. Having a per default tag for all images (like darktable|something) we can mirror the current behavior of the normal sorting. I haven't gone there because the position sorting on images table should stay faster than on tagged_images table (far less records in the first table) but I haven't run any measure.

File: src/common/collection.h (hunk 1 above)
Comment (phweyland) [reply src/common/collection.h:79]: right !

File: src/common/database.c (hunk 2 above)
Comment (phweyland) [reply src/common/database.c:1655]: Right.

**Comment by phweyland** [comment]
Updated with fixes related to your comments.

I'll remove all album attributes keeping order on tags on the next push. This should simplify the code. If not convenient we'll be able to come back to this commit.

File: src/common/image.c [hunk 7]
```diff
@@ -1811,9 +1818,13 @@ int32_t dt_image_copy_rename(const int32_t imgid, const int32_t filmid, const gc
         sqlite3_step(stmt);
         sqlite3_finalize(stmt);
         DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
-                                    "INSERT INTO main.tagged_images (imgid, tagid)"
-                                    " SELECT ?1, tagid"
-                                    " FROM main.tagged_images"
+                                    "INSERT INTO main.tagged_images (imgid, tagid, position)"
+                                    " SELECT ?1, tagid, "
+                                    // 4294967295 << 32 = 0xFFFFFFFF00000000
```
Comment (TurboGit) [inline src/common/image.c:1823]: Looks like SQlite3 supports 0xFFFFFFFF00000000 directly have you tried to replace:
`
(4294967295 << 32)`
by
`0xFFFFFFFF00000000`

Maybe clearer to the reader.

**Comment by phweyland** [comment]
In that commit I've removed all references to album.
Then tags are now order aware when in first place in collect filter.

File: src/common/image.c (hunk 7 above)
Comment (phweyland) [reply src/common/image.c:1823]: I'd read somewhere (probably not in the right place) that we could not do that. I haven't given a try....
Done.

**Comment by phweyland** [comment]
0xFFFFFFFF00000000 ! :)

**Comment by phweyland** [comment]
minor formatting plus one fix.

**Comment by TurboGit** [comment]
>     * add position to tagged_images

This part is not clear, why do we want to have the position of a tag ? This makes the tagged_images more than just a unique relation between an imgid and a tagid. So in turns this means that we can have the same tag at different position ? Again, this looks quite complex to me and could certainly hinder the maintenance!

**Comment by phweyland** [comment]
> This part is not clear, why do we want to have the position of a tag ?

It is not adding position to a tag.

This adding a position for the the couple (tagid, imagid).
Then for each tagid, images cab be ordered.

**Comment by phweyland** [comment]
> This makes the tagged_images more than just a unique relation between an imgid and a tagid.

No, this doesn't change anything to the current role of tagged_images.

Just adding the position attribute to (tagid, imgid). This position is used in collect module instead of the images table position when the user has selected a tag.
So, as we have the global ordering today, this change adds the possibility (with custom sort order) order manually the images for each tag.

To resume:
If the first collect filter is anything else than tag the images order is as before (no change).
When the first collect filter is tag,  any change on the order selection is kept along the tagid (in flags of tags table). And if the sort order is set as custom sort, the user can drag and drop the images in a different order. This order is kept in tagged_images position.

**Comment by TurboGit** [comment]
I see, I reread the corresponding issue. This is clearer to me now. I'll review and test again in the coming days with this in mind. Far more challenging that what I thought initially !

**Comment by TurboGit** [comment]
Ok, tested and working on my side. Thanks Philippe and sorry for all my questions, I have hard time to follow closely everything.

**Comment by phweyland** [comment]
> and sorry for all my questions

Pascal, you do a tremendous job ! That's absolutely normal you ask questions. For sure you have to understand what's happening. And if it is not clear, most of the time that's because the presentation or the explanations are not good enough.

Sometimes I have the impression that dt changes are going a bit fast :).

Thanks for the questions !

**Comment by Nilvus** [comment]
@phweyland: if you can, could you read this issue on french forum: https://forums.darktable.fr/showthread.php?tid=4971&pid=42180#pid42180
A user describe an issue with film roll with clones inside. And the master he use is the last one from same day this PR is merged and it's the only thing related to database as he talks about an update.

**Comment by phweyland** [comment]
> if you can, could you read this issue on french forum: https://forums.darktable.fr/showthread.php?tid=4971&pid=42180#pid42180

I've posted a question there but I don't see my message... not sure it has gone there ...
I've some duplicates in my database and I've not reproduced yet the issue.

**Comment by spaceChRiS** [comment]
> Pascal, you do a tremendous job !

Fully agree here. And @phweyland, you too! And all the other contributors. What a magnificent project this is. Since this PR is a bit my fault (k, not really a fault), a big thank you to both of you for implementing and testing! This was extremely fast, I cannot say how thankful I am. This is incredible :smile:.

Unfortunately, I cannot test as darktable no longer builds on my ubuntu 16.04. I am working on this, but it will take a couple of weeks :expressionless:.

I wonder if allowing the sorting for all tags will mess up the data base and produce a lot of unnecessary entries? May it therefore not be better to restrict this feature to tags that have the album property, as in your original implementation?

**Comment by phweyland** [comment]
> I wonder if allowing the sorting for all tags will mess up the data base and produce a lot of unnecessary entries?

**Comment by phweyland** [comment]
> I wonder if allowing the sorting for all tags will mess up the data base and produce a lot of unnecessary entries?

I started this way but it adds a lot of specific cases and is slower especially acting on db. Applying it on all tags keeps it simple. Furthermore the tag order only intervenes when tag is the first entry of collect module. Otherwise everything works as before. 
I'm looking forward to your feedback on this!

**Comment by jerrac** [comment]
I tried installing the latest stable version from the OBS repo, then the latest snapshot version from the OBS snapshot repo, then I finally compiled the master branch and installed it. Every time I edit a tag, there is no "album" checkbox.

So, how does the feature this pull request is about actually work?

**Pull request merged**